src/network/tarcap/packets_handler/handlers/dag_packets_handler.cpp

```diff
@@ -0,0 +1,230 @@
+#include "dag_packets_handler.hpp"
+
+#include "dag/dag_block_manager.hpp"
+#include "network/tarcap/packets_handler/syncing_state.hpp"
+#include "transaction_manager/transaction_manager.hpp"
+
+namespace taraxa::network::tarcap {
+
+DagPacketsHandler::DagPacketsHandler(std::shared_ptr<PeersState> peers_state,
+                                     std::shared_ptr<SyncingState> syncing_state,
+                                     std::shared_ptr<TransactionManager> trx_mgr,
+                                     std::shared_ptr<DagBlockManager> dag_blk_mgr, std::shared_ptr<DbStorage> db,
+                                     uint16_t network_min_dag_block_broadcast, uint16_t network_max_dag_block_broadcast,
+                                     const addr_t &node_addr)
+    : PacketHandler(std::move(peers_state), node_addr, "DAG_BLOCK_PH"),
+      syncing_state_(syncing_state),
+      trx_mgr_(trx_mgr),
+      dag_blk_mgr_(dag_blk_mgr),
+      db_(db),
+      urng_(std::mt19937_64(std::random_device()())),
+      network_min_dag_block_broadcast_(network_min_dag_block_broadcast),
+      network_max_dag_block_broadcast_(network_max_dag_block_broadcast) {}
+
+void DagPacketsHandler::process(const PacketData &packet_data, const dev::RLP &packet_rlp) {
+  if (packet_data.type_ == NewBlockPacket) {
+    processNewBlockPacket(packet_data, packet_rlp);
+  } else if (packet_data.type_ == NewBlockHashPacket) {
+    processNewBlockHashPacket(packet_data, packet_rlp);
+  } else if (packet_data.type_ == GetNewBlockPacket) {
+    processGetNewBlockPacket(packet_data, packet_rlp);
+  } else {
+    assert(false);
+  }
+}
+
+inline void DagPacketsHandler::processNewBlockPacket(const PacketData &packet_data, const dev::RLP &packet_rlp) {
+  // Ignore new block packets when syncing
+  if (syncing_state_->is_syncing()) return;
+
+  DagBlock block(packet_rlp[0].data().toBytes());
+  blk_hash_t const hash = block.getHash();
+
+  if (dag_blk_mgr_) {
+    if (dag_blk_mgr_->isBlockKnown(hash)) {
+      LOG(log_dg_) << "Received NewBlock " << hash.toString() << "that is already known";
+      return;
+    }
+    if (auto status = syncing_state_->checkDagBlockValidation(block); !status.first) {
+      LOG(log_wr_) << "Received NewBlock " << hash.toString() << " missing pivot or/and tips";
+      status.second.push_back(hash);
+      syncing_state_->requestBlocks(packet_data.from_node_id_, status.second,
+                                    GetBlocksPacketRequestType::MissingHashes);
+      return;
+    }
+  }
+
+  const auto transactions_count = packet_rlp.itemCount() - 1;
+  LOG(log_dg_) << "Received NewBlockPacket " << transactions_count;
+
+  std::vector<Transaction> new_transactions;
+  for (size_t i_transaction = 1; i_transaction < transactions_count + 1; i_transaction++) {
+    Transaction transaction(packet_rlp[i_transaction].data().toBytes());
```

Comment:
this "handlers separation" might be a good task to get rid of all rlp random access like here "packet_rlp[i_transaction]"

Reply:
this is not an issue actually if you access elements in ascending order it is exactly same as iterator

Reply:
hmm I did not really check but @mfrankovi suggested that when using random access it always iterates from 0 until it gets to the index you want. If true, it does not matter if it is in order or not...

Reply:
it is even noted in the header, so we should be alright
```
  /// @note if used to access items in ascending order, this is efficient.
  RLP operator[](size_t _i) const;
```

Reply:
it is more complicated, it will mark last index so in ascending order is exactly the same as iterator, if you really do random access it will iterate from start

Reply:
ah okey, good to know :)